src/watch-run.ts

```diff
@@ -37,26 +47,58 @@ export function makeWatchRun(instance: TSInstance) {
           filePath.match(constants.dtsDtsxOrDtsDtsxMapRegex) !== null &&
           filePath.match(constants.nodeModules) === null
         ) {
-          updateFile(instance, filePath);
+          promises.push(updateFile(instance, loader, loaderIndex, filePath));
         }
       }
     }
 
     // Update all the watched files from solution builder
     if (instance.solutionBuilderHost) {
       for (const filePath of instance.solutionBuilderHost.watchedFiles.keys()) {
-        updateFile(instance, filePath);
+        promises.push(updateFile(instance, loader, loaderIndex, filePath));
       }
     }
 
-    callback();
+    Promise.all(promises)
+      .then(() => callback())
+      .catch(err => callback(err));
   };
 }
 
-function updateFile(instance: TSInstance, filePath: string) {
-  updateFileWithText(
-    instance,
-    filePath,
-    nFilePath => readFile(nFilePath) || ''
-  );
+function updateFile(
+  instance: TSInstance,
+  loader: webpack.loader.LoaderContext,
+  loaderIndex: number,
+  filePath: string
+) {
+  return new Promise<void>((resolve, reject) => {
+    if (instance.rootFileNames.has(path.normalize(filePath))) {
+      loaderRunner.runLoaders(
+        {
+          resource: filePath,
+          loaders: loader.loaders
+            .slice(loaderIndex + 1)
+            .map(l => ({ loader: l.path, options: l.options })),
+          context: {},
+          readResource: fs.readFile.bind(fs)
```

Comment:
It looks like this will create a new instance of `readFile` each time `updateFile` runs. Would it be more performant to create this once and reuse it? In fact is the `bind` necessary at all?

Reply:
It may be a bit efficient to cache the result of `fs.readFile.bind(fs)` as you say.
I think `bind` is necessary, because there is no guarantee that `fs.readFile` does not use `this` internally.

... In the first place, `readResource` is not necessary, because it is optional and defaulted to `readFile`. As [@types/loader-runner](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/loader-runner/index.d.ts#L25) does not mark it as optional, I pass the default value explicitly to avoid a type error.
This may be better:
```
runLoaders(
  {
    resource: ...,
    loaders: ...
  } as loaderRunner.RunLoaderOption,
```